### DISCUSSION TITLE
WPF Command Equivalent

### DISCUSSION THREAD

**Discussion by gebodal** [AvaloniaUI/Avalonia#15963] [Q&A]
Hello all,

I’m getting a bit lost trying to understand Commands in Avalonia. The [documentation pages](https://docs.avaloniaui.net/docs/0.10.x/data-binding/binding-to-commands) go straight to talking about ReactiveUI, and I can’t work out if commands (as in, behaviours defined at the semantic level, separate from the key presses that trigger them, like in WPF) are possible using the Community Toolkit?

But I saw in [some comments](https://github.com/AvaloniaUI/Avalonia/issues/2611#issuecomment-498232434) that there are specific Paste events in Avalonia, so I’m guessing there must be some way to utilise them as the targets of `Button`s and `MenuItem`s?

Can anyone point me to some documentation on creating custom commands, and utilising existing commands, within the Community Toolkit approach? Or could someone give me some pointers, by any chance?

Specifically:
- Where do we register commands, or which object owns them?
- How do we reference them as the targets of `Button`s and `MenuItem`s?
- What built-in commands are available? (I’m particularly interested in Cut/Copy/Paste, and NavigateForward/GoForward or whatever it may be called in Avalonia)

Many thanks in advance!

**Comment by FirehawkV21** [comment]
You can do it just like WPF (when using the CommunityToolkit.MVVM):

C#
```cs
    [RelayCommand]
    private static void CancelButton() => s_abortTokenDispatch.Cancel();
```

XAML
```xaml
            <Button Name="CancelButton" Content="Cancel" Command="{Binding CancelButtonCommand}"/>
```

You may have to build, so Intellisense can see the binding. You will need to attach the desired ViewModel to the View (or window) like this:

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:App.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="640" d:DesignHeight="480" MinHeight="480" MinWidth="640" Width="640" Height="480" SizeToContent="WidthAndHeight" WindowStartupLocation="CenterScreen" CanResize="False" 
        x:Class="App.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel">
        <Panel>
            <Design.DataContext>
      <!-- This only sets the DataContext for the previewer in an IDE,
             to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
      <vm:MainWindowViewModel/>
    </Design.DataContext>
        </Panel>
```

Essentially, add the reference to the `viewmodels` namespace(`xmlns:vm="using:App.ViewModels"`) and set the DataType (`x:DataType="vm:MainWindowViewModel"`). After that, add the Design.DataContext bit so the IDE can show you the bindings.

**Reply by gebodal** [reply]
Thanks for the help!

Aha, so something like this will automatically sort out the button being enabled, then, right?

```
[RelayCommand(CanExecute = nameof(CanCancelButton))]
private void CancelButton() {
    Console.WriteLine($"Cancelled!");
}

private bool CanCancelButton() {
    return canCancel; // Some state variable
}
```

A follow-up question, if I may: Can one assign hotkeys to `RelayCommand`s in any way from Avalonia, or does that have to be done manually through `KeyDown` events? And does Avalonia have any understanding of more abstract "gestures"/"inputs" like "Navigate Forward"?

**Reply by FirehawkV21** [reply]
I usually use an `ObservableProperty` to disable/enable the button.

To add a keybind for a command:

```xaml
  <Window.KeyBindings>
    <KeyBinding Gesture="F1" Command="{Binding OpenHelpSiteCommand}"/>
    <KeyBinding Gesture="Shift+F1" Command="{Binding OpenBugReportSiteCommand}"/>
  </Window.KeyBindings>
  ```

**Comment by stevemonaco** [comment]
Can you be more specific about the problem you're trying to solve here? It doesn't seem like you're only asking about typical ViewModel commands that implement your own logic, but instead commands that work directly with View components.

WPF had `ApplicationCommands` which contains `RoutedCommand`s, but there's no equivalent for Avalonia. Avalonia exposes some gestures to trigger, provides a public method to call, and works with events at the moment.

**Reply by gebodal** [reply]
Thanks for the response!

I'm afraid I'm still a green with regards to proper MVVM architectures, so I'm probably not phrasing the question clearly enough! Also, the below is probably horribly hacky, but it works (small justification, but still).

Specifically, this is the code I'm trying to replicate in Avalonia:

Let's say I have the following commands in `MyWindow.xaml`:
```
<Window.CommandBindings>
	<!-- Built in application command -->
	<CommandBinding Command="ApplicationCommands.Save"
		Executed="SaveExecuted"
		CanExecute="SaveCanExecute"/>
	<!-- Built in navigation command -->
	<CommandBinding Command="NavigationCommands.BrowseBack"
                Executed="BrowseBackExecuted"
                CanExecute="NavigationCommandCanExecute"/>
	<!-- Fully custom command -->
	<CommandBinding Command="{x:Static local:MyWindow.GenerateCommand}"
		Executed="GenerateCommandExecuted"
		CanExecute="GenerateCommandCanExecute"/>
</Window.CommandBindings>
```

Three different types of command here: Application, Navigation, and fully custom.

In `MyWindow.xaml.cs` I then have the following to wire up the actual command functionality:
```
public static RoutedUICommand GenerateCommand = new RoutedUICommand(
	"Generate", "Generate",
	typeof(MyWindow),
	new InputGestureCollection() {
		new KeyGesture(Key.G, ModifierKeys.Control)
	});

private void GenerateCommandExecuted(object? sender, ExecutedRoutedEventArgs e) {
	// Perform generation operations
}

private void GenerateCommandCanExecute(object? sender, CanExecuteRoutedEventArgs e) {
	e.CanExecute = GenerateEnabled; // An observable property
}
```

And then back in `MyWindow.xaml` these commands are used like so:
```
<Button Command="Save">
	<Image Source="Images/Save.png" Height="16"/>
</Button>
<Button Command="{x:Static local:MyWindow.GenerateCommand}">
	<Image Source="Images/Generate.png" Height="16"/>
</Button>
```

In this setup, Ctrl-S on the window triggers the `SaveExecuted` method, and Ctrl-G triggers `GenerateCommandExecuted`, and using mouse buttons 4 & 5 trigger forward and backward navigation (assuming that `NavigationCommands.BrowseForward` is also implemented equivalently). All this without any additional `PreviewKeyDown`/`PreviewMouseButtonDown` or what-have-you listening for these inputs.

I was hoping that there was some equivalent for Avalonia that listened to the appropriate inputs and forwarded them to the appropriate commands, using whatever input-semantics magic is needed to wire this up properly (I have no idea how the semantics of the command "Navigate Forward" are implemented in Windows for WPF, let alone cross-platform).

However, based on your comment, I'm guessing Avalonia doesn't have this functionality? When you say "gestures", is this just "PointerPressed" and the like, or is there a more abstract "NavigateBack" equivalent? And if so, where is the documentation for it?

**Reply by stevemonaco** [reply]
So Avalonia doesn't have `RoutedCommand` / `RoutedUICommand`, `ApplicationCommands`, or `NavigationCommands`. I'm going to be opinionated and say those all kind of suck anyways, in addition to WPF's poor navigation support (Avalonia lacks there, too, but having no support is better than having bad support when DIY navigation is easy when you know what you're writing).

For custom, you can use `KeyBinding`. The input event will bubble up rather than a command.

The gestures I'm talking about is the hotkey used to execute the action. For example, see `TextBox`:

https://github.com/AvaloniaUI/Avalonia/blob/82e3613939af5451597d3c19f3a48e0a8ba888f3/src/Avalonia.Controls/TextBox.cs#L35

You'll see that it exposes things like `Cut`, `CanCut`, and `CuttingToClipboardEvent`. So in Avalonia, you would subscribe to `CuttingToClipboard`, handle it in code-behind, and forward the action to your ViewModel or send a message. Same goes for navigation, except there aren't any navigation events but you use less specific events like selection changed, clicked, etc.

**Comment by stevemonaco** [comment]
I didn't notice initially, but the docs you're linking are the 0.10.x docs instead of the Avalonia 11 docs. There's still a fair bit of ReactiveUI sections in the v11 docs, but it's a bit less used.

**Comment by maxkatz6** [comment]
If you need command manager, you can try to use an experimental semi-official port https://github.com/AvaloniaUI/Avalonia.Labs/tree/main/src/Avalonia.Labs.CommandManager

**Reply by djrecipe** [reply]
I will piggyback off this thread to ask a question: how to force re-evaluation of `CanExecute` methods in Avalonia. In other words, how to replicate `CommandManager.InvalidateRequerySuggested();` from WPF, to help disable/reenable buttons?

**Reply by jm-abellan** [reply]
Can't you just call NotifyCanExecuteChanged() for your ICommand?
Just do it in UI thread if you are out of it when require notification update

**Reply by djrecipe** [reply]
Doh! Yes, thank you very much

**Comment by DamirLisak** [comment]
The fact that the copy/paste functions cannot be called via a key is not due to the suggestion to use Avalonia.Labs.Input.CommandManager with CommandBindings or ReactiveCommand, but rather to an insufficient implementation of the KeyBinding and HotKeyManager classes.

The problem is that in line 40, `args.Handled` is set to `true`, and therefore the RouteEvent is not forwarded to the TextBox, or the method `protected override void OnKeyDown(KeyEventArgs e))` is called:

![KeyBinding](https://github.com/user-attachments/assets/265f5178-3ae9-490f-981f-8e83a6c83224)

The reason the program ends up in an `if` block is due to the insufficient implementation of the `CanExecute()` method of the HotKeyManager, which either calls `ICommand.CanExecute` (regardless of whether it's ReactiveCommand or Avalonia.Labs.Input.RoutedCommand) if it's set on the button, or, if it's not set, the fallback mechanism in line 50 is used by checking whether the key can be pressed:

![HotKeyManager](https://github.com/user-attachments/assets/3801b54b-91e9-467f-9081-c9e22cde5099)

Both cases are only correct if you actually want your own `CanExecute` method to be called and not Avalonia's built-in functionality of the textbox's cut-and-paste mechanism. There is no workaround here, because if you were to return "false" in your own `CanExecute` method to enable key binding, the key would be disabled, which is undesirable.

This implementation has been active since commit #d62381a, and perhaps @workgroupengineering (Guiseppe Lippolis) can shed some light on why it was implemented this way.
https://github.com/AvaloniaUI/Avalonia/commit/d62381a0b218e063fcfb7b725130e5fead597428

**Reply by workgroupengineering** [reply]
Sorry, I'm not very active on Avalonia anymore, and I wrote that code a long time ago. I'm not quite sure what you're trying to achieve.

**Reply by DamirLisak** [reply]
What I want to achieve is the same as @gebodal. I don't want to implement a separate command handler for the standard key gestures (Ctrl-C, Ctrl-V, Ctrl-X,...) that are defined in the Avalonias `PlatformHotkeyConfiguration` class and used in controls such as the Avalonias `TextBox`. This is especially important if, for example, I have buttons in a Ribbon control where I assign these gestures to the HotKey property of a Button. 

In short, I want the user to be able to press a button and execute this "built-in-commands" without having to open the context menu on the TextBox or press the keyboard shortcut.

In WPF, you could previously simply create a CommandBinding on the `System.Windows.Input.ApplicationCommands`, and it worked "plug and play". The equivalent of the `InputGestures` property in the `RoutedCommand` class should be the `Gestures` property in Avalonia. As mentioned above, the Avalonia.Labs implementation of the CommandManager is not entirely compliant with WPF. That's probably why @maxkatz6 called it an "experimental semi-official port". In some configurations, the CommandManager doesn't catch all routed events that should actually be intercepted via CommandBinding. Therefore, we prefer ReactiveCommands, as recommended by Avalonia.

To solve this problem, the `CanExecute` property of the `HotkeyCommandWrapper` would need to be modified so that if the `HotKey` property is set but the `ICommand Command` property is not, it returns `false`. The program should not enter the `if` block in lines 48-51 where `clickable.IsEffectivelyEnabled` is returned, as the button should be pressable. However, since I don't know your intention with this change, which cases it should cover, or what side effects such a change might have, I've referred you to the question in the hope that you can explain your reasoning.
